src/charm.py

```diff
@@ -9,76 +9,181 @@
 """
 
 import logging
+from typing import Dict
 
-from charms.data_platform_libs.v0.database_requires import (
+from charms.data_platform_libs.v0.data_interfaces import (
     DatabaseCreatedEvent,
     DatabaseRequires,
+    KafkaRequires,
+    TopicCreatedEvent,
 )
 from ops.charm import ActionEvent, CharmBase
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus
 
 logger = logging.getLogger(__name__)
 
+KAFKA_EXTRA_USER_ROLES = "consumer,producer"
+
 
 class IntegratorCharm(CharmBase):
     """Integrator charm that connects to database charms."""
 
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.database = self.model.config.get("database", "")
+        self.database = self.model.config.get("database-name", "")
+        self.topic = self.model.config.get("topic-name", "")
+        self.extra_user_roles = self.model.config.get("extra-user-roles", "")
 
         self.framework.observe(self.on.get_credentials_action, self._on_get_credentials_action)
         self.framework.observe(self.on.start, self._on_config_changed)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
 
         # MySQL
-        self.mysql = DatabaseRequires(self, relation_name="mysql", database_name=self.database)
+        self.mysql = DatabaseRequires(
+            self,
+            relation_name="mysql",
+            database_name=self.database,
+            extra_user_roles=self.extra_user_roles,
+        )
         self.framework.observe(self.mysql.on.database_created, self._on_database_created)
+        self.framework.observe(self.on["mysql"].relation_broken, self._on_database_relation_broken)
 
         # PostgreSQL
         self.postgresql = DatabaseRequires(
             self, relation_name="postgresql", database_name=self.database
         )
         self.framework.observe(self.postgresql.on.database_created, self._on_database_created)
+        self.framework.observe(
+            self.on["postgresql"].relation_broken, self._on_database_relation_broken
+        )
 
         # MongoDB
         self.mongodb = DatabaseRequires(self, relation_name="mongodb", database_name=self.database)
         self.framework.observe(self.mongodb.on.database_created, self._on_database_created)
+        self.framework.observe(
+            self.on["mongodb"].relation_broken, self._on_database_relation_broken
+        )
+
+        kafka_user_roles = KAFKA_EXTRA_USER_ROLES + self.extra_user_roles
```

Comment:
Is it by design that the KAFKA_EXTRA_USER_ROLES cannot be overridden? WHy is it so?

I mean, could I get a user with "producer" permission only?

Reply:
I removed that